plat/common/paging.c

```diff
@@ -0,0 +1,1320 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/* API based on prototype for virtual memory by:
+ * Stefan Teodorescu <stefanl.teodorescu@gmail.com>
+ * See https://github.com/unikraft/unikraft/pull/247
+ */
+
+#include <string.h>
+#include <errno.h>
+
+#include <uk/config.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/lcpu.h>
+#include <uk/essentials.h>
+#include <uk/assert.h>
+#include <uk/print.h>
+#include <uk/plat/paging.h>
+#include <uk/falloc.h>
+
+#define __PLAT_CMN_ARCH_PAGING_H__
+#ifdef CONFIG_ARCH_X86_64
+#include <x86/paging.h>
+#else
+#error "Architecture not supported by paging API"
+#endif
+
+/* Forward declarations */
+static inline int pg_pt_alloc(struct uk_pagetable *pt, __vaddr_t *pt_vaddr,
+			      __paddr_t *pt_paddr, unsigned int level);
+
+static inline void pg_pt_free(struct uk_pagetable *pt, __vaddr_t pt_vaddr,
+			      unsigned int level);
+
+static int pg_page_unmap(struct uk_pagetable *pt, __vaddr_t pt_vaddr,
+			 unsigned int level, __vaddr_t vaddr, __sz len,
+			 unsigned long flags);
+
+#ifdef CONFIG_PAGING_STATS
+#define PAGE_FLAG_INTERN_STATS_KEEP	0x80000000 /* Don't update stats */
+#endif /* CONFIG_PAGING_STATS */
+
+/* The largest level for which PAGE_Lx_HAS() returns a non-zero value, that is
+ * the largest level which can map a page.
+ */
+static unsigned int pg_page_largest_level = PAGE_LEVEL;
+
+/*
+ * Pointer to currently active page table.
+ * TODO: With SMP support, this should move to a CPU-local variable or we
```

Comment:
How about this? Since this is planned to be merged before SMP support, maybe it would worth to create an issue in the backlog?

Reply:
Yup, created a corresponding issue #426 👍🏻